### PR TITLE
SMR: Frontend support for Dedicated Servers (WIP)

### PR DESCRIPTION
**Pull request by porisius**
Functional complete, design/formatting work remains.

### CODE REVIEW

File: src/lib/components/sml-versions/SMLVersionForm.svelte [hunk 1]
```diff
@@ -30,6 +35,30 @@
...
 
+  const add = () => {
+    $data.arch = $data.arch.concat({ id: '', SMLVersionLinkID: '', platform: '', link: '' });
+  };
+
+  const remove = (i) => () => {
+    $data.arch = $data.arch.filter((_, index) => index != i);
+  };
+
+  function isAdd(index) {
```
Comment (Vilsol) [inline src/lib/components/sml-versions/SMLVersionForm.svelte:46]: Please rename to `canAdd`

File: src/lib/components/sml-versions/SMLVersionForm.svelte [hunk 2]
```diff
@@ -30,6 +35,30 @@
...
+    if (index != $data.arch.length - 1) {
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  function isDel() {
```
Comment (Vilsol) [inline src/lib/components/sml-versions/SMLVersionForm.svelte:54]: Please rename to `canDel`

File: src/lib/components/sml-versions/SMLVersionForm.svelte [hunk 3]
```diff
@@ -30,6 +35,30 @@
...
+  const remove = (i) => () => {
+    $data.arch = $data.arch.filter((_, index) => index != i);
+  };
+
+  function isAdd(index) {
+    if (index != $data.arch.length - 1) {
+      return true;
+    } else {
+      return false;
+    }
```
Comment (Vilsol) [inline src/lib/components/sml-versions/SMLVersionForm.svelte:51]: Please change the function to just `return index != $data.arch.length - 1`

File: src/lib/components/sml-versions/SMLVersionForm.svelte [hunk 4]
```diff
@@ -30,6 +35,30 @@
...
+  }
+
+  function isDel() {
+    if ($data.arch.length == 1) {
+      return true;
+    } else {
+      return false;
+    }
```
Comment (Vilsol) [inline src/lib/components/sml-versions/SMLVersionForm.svelte:59]: Please change the function to just `$data.arch.length == 1`

File: src/lib/models/sml-versions.ts [hunk 5]
```diff
@@ -1,25 +1,41 @@
 import * as zod from 'zod';
 import type { VersionStabilities } from '$lib/generated';
 
+export type SMLLinkData = {
```
Comment (Vilsol) [inline src/lib/models/sml-versions.ts:4]: Please change to `SMLArchData`

File: src/lib/models/sml-versions.ts [hunk 6]
```diff
@@ -1,25 +1,41 @@
...
   link: string;
+  arch: SMLLinkData[];
   changelog: string;
   date: string;
 };
 
 const versionRegex =
   /^(<=|<|>|>=|\^)?(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$/;
 
+export const smlLinksSchema = zod.object({
```
Comment (Vilsol) [inline src/lib/models/sml-versions.ts:25]: Please rename to `smlArchsSchema`

File: src/lib/models/versions.ts [hunk 7]
```diff
@@ -24,6 +25,15 @@ export type VersionMetadata = {
   objects: string[];
 };
 
+export type ModLinkData = {
```
Comment (Vilsol) [inline src/lib/models/versions.ts:28]: Please rename to `ModArchData`

File: src/lib/utils/formatting.ts [hunk 8]
```diff
@@ -23,3 +23,16 @@ export const prettyBytes = (bytes: number, decimals = 2) => {
...
 };
+
+export const prettyArch = (platform: string) => {
+  switch (platform) {
+    case 'WindowsNoEditor':
+      return 'Client';
+    case 'WindowsServer':
+      return 'Windows Server';
+    case 'LinuxServer':
+      return 'LinuxServer';
```
Comment (Vilsol) [inline src/lib/utils/formatting.ts:34]: Please change to `Linux Server`

File: src/lib/utils/formatting.ts [hunk 9]
```diff
@@ -23,3 +23,16 @@ export const prettyBytes = (bytes: number, decimals = 2) => {
 
   return parseFloat((bytes / Math.pow(k, i)).toFixed(dm)) + ' ' + sizes[i];
 };
+
+export const prettyArch = (platform: string) => {
+  switch (platform) {
+    case 'WindowsNoEditor':
+      return 'Client';
```
Comment (Vilsol) [inline src/lib/utils/formatting.ts:30]: Please change to `Windows Client`

**Review by Vilsol — CHANGES_REQUESTED** [review]

File: .husky/.gitignore [hunk 10]
```diff
@@ -1 +0,0 @@
-_
```
[the remark this answers is not in the document]
Comment (porisius) [reply .husky/.gitignore]: https://github.com/satisfactorymodding/smr-frontend/pull/47/commits/b2dd492b7059dc21bbcccec615cbd7a73e1c000a

File: .husky/commit-msg [hunk 11]
```diff
@@ -1,4 +0,0 @@
-#!/bin/sh
```
[the remark this answers is not in the document]
Comment (porisius) [reply .husky/commit-msg]: https://github.com/satisfactorymodding/smr-frontend/pull/47/commits/b2dd492b7059dc21bbcccec615cbd7a73e1c000a

File: .husky/pre-commit [hunk 12]
```diff
@@ -1,4 +0,0 @@
-#!/bin/sh
```
[the remark this answers is not in the document]
Comment (porisius) [reply .husky/pre-commit:1]: https://github.com/satisfactorymodding/smr-frontend/pull/47/commits/b2dd492b7059dc21bbcccec615cbd7a73e1c000a

File: src/lib/utils/formatting.ts (hunk 9 above)
Comment (porisius) [reply src/lib/utils/formatting.ts]: https://github.com/satisfactorymodding/smr-frontend/pull/47/commits/9c0893e0353c97b347f9f0152ec590c614d255ac

File: src/lib/utils/formatting.ts (hunk 8 above)
Comment (porisius) [reply src/lib/utils/formatting.ts]: https://github.com/satisfactorymodding/smr-frontend/pull/47/commits/9c0893e0353c97b347f9f0152ec590c614d255ac

File: src/lib/models/versions.ts (hunk 7 above)
Comment (porisius) [reply src/lib/models/versions.ts]: https://github.com/satisfactorymodding/smr-frontend/pull/47/commits/9c0893e0353c97b347f9f0152ec590c614d255ac

File: src/lib/models/sml-versions.ts (hunk 6 above)
Comment (porisius) [reply src/lib/models/sml-versions.ts:25]: https://github.com/satisfactorymodding/smr-frontend/pull/47/commits/9c0893e0353c97b347f9f0152ec590c614d255ac

File: src/lib/models/sml-versions.ts (hunk 5 above)
Comment (porisius) [reply src/lib/models/sml-versions.ts]: https://github.com/satisfactorymodding/smr-frontend/pull/47/commits/9c0893e0353c97b347f9f0152ec590c614d255ac

File: src/lib/components/sml-versions/SMLVersionForm.svelte (hunk 2 above)
Comment (porisius) [reply src/lib/components/sml-versions/SMLVersionForm.svelte]: https://github.com/satisfactorymodding/smr-frontend/pull/47/commits/9c0893e0353c97b347f9f0152ec590c614d255ac

I removed it and isAdd, but it still functions as needed.

File: src/lib/components/sml-versions/SMLVersionForm.svelte (hunk 1 above)
Comment (porisius) [reply src/lib/components/sml-versions/SMLVersionForm.svelte]: https://github.com/satisfactorymodding/smr-frontend/pull/47/commits/9c0893e0353c97b347f9f0152ec590c614d255ac

I removed it and isAdd, but it still functions as needed.

File: src/lib/components/sml-versions/SMLVersionForm.svelte (hunk 3 above)
Comment (porisius) [reply src/lib/components/sml-versions/SMLVersionForm.svelte]: https://github.com/satisfactorymodding/smr-frontend/pull/47/commits/9c0893e0353c97b347f9f0152ec590c614d255ac

I removed it and isAdd, but it still functions as needed.

File: src/lib/components/sml-versions/SMLVersionForm.svelte (hunk 4 above)
Comment (porisius) [reply src/lib/components/sml-versions/SMLVersionForm.svelte]: https://github.com/satisfactorymodding/smr-frontend/pull/47/commits/9c0893e0353c97b347f9f0152ec590c614d255ac

I removed it and isAdd, but it still functions as needed.

**Review by Vilsol — APPROVED** [review]

**Review by Vilsol — APPROVED** [review]

**Review by Vilsol — APPROVED** [review]

**Pull request merged**